solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java

```diff
@@ -0,0 +1,328 @@
+  private final TimeoutCallback timeoutCallback;
+  private volatile Future<?> timeoutTask = null;
+
+  /**
+   * A ShardRequestActor that times out all pending requests if all of them are from slow nodes.
+   * Each instance of this class is tied to a single ShardRequest instance, which can be
+   * retried/submitted to many shard urls.
+   *
+   * @param timeout timeout in milliseconds, if all pending requests remain are from slow nodes, a
+   *     task to cancelled all pending requests will be scheduled with this timeout as delay
+   * @param dryRun whether to actually cancel the pending requests or just log the cancellation
+   * @param slowNodes set of slow nodes, this should not change during the lifetime of this actor
+   * @param timeoutCallback callback to be invoked when timeout is reached, whether the pending
+   *     requests are cancelled due to dryRun mode
+   */
+  SlowNodeTimeoutActor(
+      long timeout, boolean dryRun, Set<String> slowNodes, TimeoutCallback timeoutCallback) {
+    this.timeout = timeout;
+    this.dryRun = dryRun;
+    this.slowNodes = slowNodes;
+    this.timeoutCallback = timeoutCallback;
+  }
+
+  @Override
+  public void onRequestSubmitted(List<String> shardUrls, Future<?> future) {
+    pendingFutures.add(future);
+
+    // This is a shardRequest being submitted to a single shard. However with multiple replicas,
+    // there will be multiple shard URLs.
+    // if any of those replicas are from a known slow node, then we will NOT count this pending
+    // future as a "fast node future". Therefore, we assume this future can be potentially slow
+    // and subjected to timeout.
+    if (!hasSlowNodeShardUrl(shardUrls)) {
+      pendingFutureCountFromFastNode.incrementAndGet();
+    }
+  }
+
+  private boolean hasSlowNodeShardUrl(List<String> shardUrls) {
+    for (String shardUrl : shardUrls) {
+      if (slowNodes.contains(Util.getNode(shardUrl))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public synchronized boolean onRequestCompleted(
+      String selectedShardUrl, List<String> shardUrls, Future<?> future, long timeElapsed) {
+    pendingFutures.remove(future);
+    if (pendingFutures.isEmpty()) { // every submitted futures are processed
+      if (timeoutTask != null) {
+        timeoutTask.cancel(true);
+      }
+      return true;
+    }
+
+    if (!hasSlowNodeShardUrl(shardUrls)) {
+      pendingFutureCountFromFastNode.decrementAndGet();
+    }
+
+    if (timeoutTask == null
+        && pendingFutureCountFromFastNode.get()
+            <= 0) { // all pending reqs are from slow nodes, start a timer to possibly timeout
+      ExecutorService executorService = null;
+      try {
+        executorService =
+            ExecutorUtil.newMDCAwareSingleThreadExecutor(
+                new SolrNamedThreadFactory("SlowNodeTimeout"));
+        timeoutTask =
+            executorService.submit(
```

Comment:
This code I'm not following. Should be part of ` onRequestSubmitted` function, and then it should watch response from slow nodes? Maybe we need to chat on this

Reply:
The code uses `onRequestCompleted` to achieve a more "adaptive" wait time. 

For example, if we have a timeout set as 5 secs:
1. For a really cheap query. We will give slow node 5 secs to finish. Since all the other nodes complete quickly (at almost no time at all), and then this will get triggered and wait for 5 secs
2. For an expensive query. Say all the other nodes took 30 secs to finish, then we will wait for additional 5 sec for the slow node, hence we will give slow node 35 secs to finish.

If we start the timer at `onRequestSubmitted`, we would not be able to tell how expensive the query will be, hence we cannot assign a wait time dynamically. We could risk timing out the request on a node just because the query itself is expensive and not because of such node being slow.